packages/shared/src/components/PostOptionsReadingHistoryMenu.tsx

```diff
@@ -0,0 +1,148 @@
+import React, { ReactElement, useContext } from 'react';
+import { Item } from '@dailydotdev/react-contexify';
+import dynamic from 'next/dynamic';
+import { QueryClient, QueryKey, useQueryClient } from 'react-query';
+import { ReadHistoryPost } from '../graphql/posts';
+import ShareIcon from '../../icons/share.svg';
+import BookmarkIcon from '../../icons/bookmark.svg';
+import BookmarkIconFilled from '../../icons/filled/bookmark_filled.svg';
+import XIcon from '../../icons/x.svg';
+import useBookmarkPost from '../hooks/useBookmarkPost';
+import AuthContext from '../contexts/AuthContext';
+import { ReadHistoryInfiniteData } from '../hooks/useInfiniteReadingHistory';
+import { useCopyLink } from '../hooks/useCopyLink';
+
+const PortalMenu = dynamic(() => import('./fields/PortalMenu'), {
+  ssr: false,
+});
+
+export type PostOptionsReadingHistoryMenuProps = {
+  post: ReadHistoryPost;
+  onHidden?: () => unknown;
+  displayCopiedMessageFunc?: () => void;
+  onHide?: (postId: string) => unknown;
+};
+
+const MenuIcon = ({ Icon }) => {
+  return <Icon className="mr-2 text-2xl" />;
+};
+
+/* eslint-disable no-param-reassign */
+const updateReadingHistoryPost =
+  (
+    queryClient: QueryClient,
+    historyQueryKey: QueryKey,
+    update: (oldPost: ReadHistoryPost) => Partial<ReadHistoryPost>,
+  ): ((args: { id: string }) => Promise<() => void>) =>
+  async ({ id }) => {
+    const oldReadingHistory =
+      queryClient.getQueryData<ReadHistoryInfiniteData>(historyQueryKey);
+    const newItems = oldReadingHistory.pages.flatMap(({ readHistory }) => {
+      const modified = readHistory.edges.map(({ node }) => {
+        if (node.post.id === id) {
+          node.post = {
+            ...node.post,
+            ...update(node.post),
+          };
+        }
+        return { node };
+      });
+      readHistory.edges = modified;
+      return {
+        readHistory,
+      };
+    });
```

Comment:
We have access to the indexes outside, we can definitely pass them down as props to the actual component that uses this function. Then we can just directly find the element to update than looping through all the elements. What do you think? You can check sample usage in the `useReadingHistory`.

Reply:
Maybe we can have a look at it together?

Reply:
Whenever you are free! 🚀